crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1023,18 +1031,104 @@ impl<'db> TypeInferenceBuilder<'db> {
         with_item: &ast::WithItem,
         definition: Definition<'db>,
     ) {
-        let expression = self.index.expression(&with_item.context_expr);
-        let result = infer_expression_types(self.db, expression);
-        self.extend(result);
+        let context_expr = self.index.expression(&with_item.context_expr);
+        self.extend(infer_expression_types(self.db, context_expr));
 
-        // TODO(dhruvmanila): The correct type inference here is the return type of the __enter__
-        // method of the context manager.
-        let context_expr_ty = self.expression_ty(&with_item.context_expr);
+        let target_ty = self.infer_context_expression(
+            &with_item.context_expr,
+            self.expression_ty(&with_item.context_expr),
+        );
 
         self.types
             .expressions
-            .insert(target.scoped_ast_id(self.db, self.scope()), context_expr_ty);
-        self.add_binding(target.into(), definition, context_expr_ty);
+            .insert(target.scoped_ast_id(self.db, self.scope()), target_ty);
+        self.add_binding(target.into(), definition, target_ty);
+    }
+
+    /// Infers the type of a context expression (`with expr`) and returns the target's type
+    ///
+    /// Returns [`Type::Unknown`] if the context expression doesn't implement the context manager protocol.
+    ///
+    /// ## Terminology
+    /// See [PEP343](https://peps.python.org/pep-0343/#standard-terminology).
+    fn infer_context_expression(
+        &mut self,
+        context_expression: &ast::Expr,
+        context_expression_ty: Type<'db>,
+    ) -> Type<'db> {
+        let context_manager_ty = context_expression_ty.to_meta_type(self.db);
+
+        let enter_ty = context_manager_ty.member(self.db, "__enter__");
+        let exit_ty = context_manager_ty.member(self.db, "__exit__");
+
+        // TODO: The checks here should be simplified to checking if context manager implements the `contextlib.AbstractContextManager` protocol.
```

Comment:
I don't actually know if simplification here is possible because simply checking if `context_manager_ty` is a subtype of `AbstractContextManager` won't give us the return type of the `__enter__` method. I think we do actually need to manually call the `__enter__` method somehow, and once we've done that we then _have_ to manually consider the case where `__enter__` isn't actually callable... it's hard for me to see what would actually be simpler once we have generalised support for `typing.Protocol`

Well, one thing we could do is to just check that the `__exit__` method is consistent with the `__exit__` method on `AbstractContextManager`, since `__exit__` method definitions can get pretty darn complicated. But we're not checking the call signature of `__exit__` at all right now, so this would be "fending off future complexity" rather than reducing any of the complexity that's already here

Reply:
I think what would become reusable is testing if the context manager has the right shape. Once we know that, we can call `enter` (which should never fail at this point)

Reply:
I think maybe what Alex is implying is that it could be kind of wasteful/unnecessary to check `__enter__` against the protocol, when simply calling it and seeing if the call succeeds or fails is actually sufficient. But this is a question that we can leave for later, I think the comment is reasonable as-is. We can always change our mind when actually implementing a TODO if it turns out not to make sense to to do it that way.

Reply:
whatever the case, this is a very long line ;)

```suggestion
        // TODO: The checks here should be simplified to checking
        // if context manager implements the `contextlib.AbstractContextManager` protocol.
```